tech-specs/src/explore/design/ibc.md

```diff
@@ -0,0 +1,400 @@
+                            //   - Refer to `ibc-rs::ics02_client::client_def::verify_upgrade_and_update_state()`
+                        }
+                    }
+                }
+            }
+
+            // connection
+            if key.segments.contains(&StringSeg("connections".to_owned())) {
+                // Check the connection state change
+                //   - none => INIT, none => TRYOPEN, INIT => OPEN, or TRYOPEN => OPEN
+                let state_change = check_connection_state(connection_id);
+                if state_change.is_initiated() {
+                    // "ConnectionOpenInit"
+                    // Assert that the corresponding client exists
+                } else {
+                    // Assert that the version is compatible
+
+                    // Verify the proofs to check the counterpart ledger's state is expected
+                    //   - The state can be inferred from the own connection state change
+                    //   - Use `ibc-rs::ics03_connection::handler::verify::verify_proofs()`
+                }
+            }
+
+            // channel handshake or closing
+            if key.segments.contains(&StringSeg("channelEnds".to_owned())) {
+                // Assert that the port is owend
+                // Check the channel state change
+                //   - none => INIT, none => TRYOPEN, INIT => OPEN, TRYOPEN => OPEN, or OPEN => CLOSED
+                let state_change = check_channel_state(channel_id);
+                if state_change.is_created() {
+                    // none => INIT
+                    // Nothing to do
+                    continue;
+                } else if state_change.is_closed() {
+                    // OPEN => CLOSED
+                    // Assert that the version is compatible
+
+                    match get_proofs(tx_data) {
+                        Some(proofs) => {
+                            // "ChanCloseConfirm"
+                            // Verify the proofs to check the counterpart ledger's state is expected
+                            //   - To check that the channel has been closed
+                            //   - Use `ibc-rs::ics04_connection::handler::verify::verify_channel_proofs()`
+                        }
+                        None => {
+                            // "ChanCloseInit"
+                            // Assert that the channel state is changed from `OPEN` to `CLOSED`
+                        }
+                    }
+                } else {
+                    // Verify the proof to check the counterpart ledger's state is expected
+                    //   - The state can be inferred from the own channel state change
+                    //   - Use `ibc-rs::ics04_connection::handler::verify::verify_channel_proofs()`
+                }
+            }
+
+            // send a packet
+            if key.segments.contains(&StringSeg("nextSequenceSend".to_owned())) {
+                match get_packet(tx_data) {
```

Comment:
I'm fixing this part.
In this case, a packet shouldn't be in the `tx_data` because it is made in a transaction. And, the packet doesn't have to be persistent.
I think it needs to be stored in a temporary space or a mechanism to pass from the transaction to the VP.

Reply:
@yito88 and I talked about this briefly and one way would be to allow a tx to write temporary arbitrary bytes, which could then be passed to the triggered VPs as an additional argument. After the VPs call, we can drop this structure. Perhaps there will be other scenarios in which we could use this.

@cwgoes what do you think?

Reply:
Hmm, that's alright I suppose, but why can't the VP just check that a commitment to the packet with expected fields was written?

Reply:
Thanks, I reconsidered that.
The VP can check the commitment.
I thought that timeout height and timestamp in the packet have to be checked in the VP. But we can get them from `tx_data` because the data should be set in `tx_data`.
I'll fix them.